### PR TITLE
Implemented staggered scheme in single component transport process

### PR DESCRIPTION
**Pull request by renchao-lu**
As titled.

All the benchmarks in the folder _SimpleSynthetics_ have been reproduced in this pull request.

To-do list:
- [x] Remove coding for PETSc;
- [x] Assign more strict tolerance for vtkdiff;
- [x] Improve naming of variables;

### CODE REVIEW

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 1]
```diff
@@ -358,27 +373,328 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+                MaterialLib::Fluid::FluidPropertyType::Density,
+                vars,
+                MaterialLib::Fluid::PropertyVariableType::p);
+
+            const double drho_dC = _process_data.fluid_properties->getdValue(
+                MaterialLib::Fluid::FluidPropertyType::Density,
+                vars,
+                MaterialLib::Fluid::PropertyVariableType::C);
+            double const velocity_magnitude = velocity.norm();
+            GlobalDimMatrixType const hydrodynamic_dispersion =
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: The definition repeats what for the monolithic scheme:
https://github.com/renchao-lu/ogs6/blob/MultiComponentTransportProcess_SS/ProcessLib/ComponentTransport/ComponentTransportFEM.h#L328

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 2]
```diff
@@ -358,27 +373,328 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+            // coupling term
+            {
+                double p0_int_pt = 0.0;
+                NumLib::shapeFunctionInterpolate(local_p0, N, p0_int_pt);
+
+                local_K.noalias() +=
+                    (-dNdx.transpose() * velocity * density * N +
+                     dNdx.transpose() * density * hydrodynamic_dispersion *
+                         dNdx +
+                     N.transpose() * decay_rate * porosity *
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: You can put these two calculations together as
```
                     N.transpose() * ( decay_rate * porosity *
                         retardation_factor * density +
                        porosity * retardation_factor * drho_dp *
                         (p_int_pt - p0_int_pt) * w / dt ) *N) 
```
to save matrix computing operation.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 3]
```diff
@@ -358,27 +373,328 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
-            &local_x[concentration_index], concentration_size);
-        auto const p_nodal_values = Eigen::Map<const NodalVectorType>(
-            &local_x[pressure_index], pressure_size);
+        if (_coupled_solutions == nullptr)  // monolithic scheme
+        {
+            auto const local_x = current_solution.get(indices);
+
+            auto const local_p = Eigen::Map<const NodalVectorType>(
+                &local_x[pressure_index], pressure_size);
+            auto concentration_index = 1 * concentration_size;
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: `1 * concentration_size`?

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 4]
```diff
@@ -358,27 +373,328 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+        {
+            std::vector<std::vector<GlobalIndexType>>
+                indices_of_all_coupled_processes(
+                    _coupled_solutions->coupled_xs.size(), indices);
+
+            auto const local_xs = getCurrentLocalSolutions(
+                *(this->_coupled_solutions), indices_of_all_coupled_processes);
+
+            auto const local_p = MathLib::toVector(local_xs[hydraulic_process_id]);
+            auto transport_process_id = 1;
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: const int

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 5]
```diff
@@ -443,6 +759,33 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
     Eigen::Vector3d getFlux(MathLib::Point3d const& pnt_local_coords,
                             double const t,
                             std::vector<double> const& local_x) const override
+    {
+        auto const local_p = Eigen::Map<const NodalVectorType>(
+            &local_x[pressure_index], pressure_size);
+        auto concentration_index = 1 * concentration_size;
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: `1 * concentration_size`?

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 6]
```diff
@@ -443,6 +759,33 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+        return calculateFlux(pnt_local_coords, t, local_p, local_C);
+    }
+
+    Eigen::Vector3d getFlux(
+        MathLib::Point3d const& pnt_local_coords,
+        double const t,
+        std::vector<std::vector<double>> const& local_xs) const override
+    {
+        auto const local_p = MathLib::toVector(local_xs[hydraulic_process_id]);
+        auto transport_process_id = 1;
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: const int

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 7]
```diff
@@ -358,27 +373,328 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+    void assembleHydraulicEquation(double const t,
+                                   std::vector<double>& local_M_data,
+                                   std::vector<double>& local_K_data,
+                                   std::vector<double>& local_b_data,
+                                   LocalCoupledSolutions const& coupled_xs)
+    {
+        auto local_p = Eigen::Map<const NodalVectorType>(
+            coupled_xs.local_coupled_xs[hydraulic_process_id].data(),
+            pressure_size);
+        auto transport_process_id = 1;
```
[the remark this answers is not in the document]
Comment (wenqing) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: const. The same for the similar others.

[the remark this answers is not in the document]
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 2 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Changed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 4 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 6 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 3 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: changed to `auto const concentration_index = 1 * ShapeFunction::NPOINTS;`

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 5 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Changed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 1 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: For sure somewhat code duplication. So far I have no good idea on how to prevent duplication of this kind.

Comment (wenqing) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: You can introduce a member function including this part together with lines 566-571, 
```

```

Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Done.

File: ProcessLib/ComponentTransport/ComponentTransportProcessData.h [hunk 8]
```diff
@@ -66,6 +62,31 @@ struct ComponentTransportProcessData
     //! Assignments are not needed.
     void operator=(ComponentTransportProcessData&&) = delete;
 
+    inline std::map<std::string, double> const getTransportProperties(
+        std::size_t element_id, std::string const& component_name) const
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportProcessData.h]: const

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 9]
```diff
@@ -358,27 +360,323 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+
+        GlobalDimMatrixType const& I(
+            GlobalDimMatrixType::Identity(GlobalDim, GlobalDim));
+
+        auto const& transport_properties = _process_data.getTransportProperties(
+            _element.getID(),
+            _process_variables[transport_process_id][0].get().getName());
+
+        auto const& solute_dispersivity_transverse =
+            transport_properties.at("solute_dispersivity_transverse");
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: don't use `at`;

**Comment by endJunction** [comment]
@renchao-lu Please rebase. Reason is the changed lfs server.

**Comment by renchao-lu** [comment]
Rebased. @endJunction

**Comment by endJunction** [comment]
Why are the ctests' tolerances mostly four orders of magnitude more relaxed?

**Comment by endJunction** [comment]
The changes in the time-loop go in own PR.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 10]
```diff
@@ -358,27 +360,323 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+            _integration_method.getNumberOfPoints();
+
+        SpatialPosition pos;
+        pos.setElementID(_element.getID());
+
+        auto const& b = _process_data.specific_body_force;
+
+        MaterialLib::Fluid::FluidProperty::ArrayType vars;
+
+        for (std::size_t ip(0); ip < n_integration_points; ++ip)
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: should be `unsigned`, same as `n_integration_points`, no need to increase the integer's size.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 11]
```diff
@@ -358,27 +360,323 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+        auto const& b = _process_data.specific_body_force;
+
+        MaterialLib::Fluid::FluidProperty::ArrayType vars;
+
+        GlobalDimMatrixType const& I(
+            GlobalDimMatrixType::Identity(GlobalDim, GlobalDim));
+
+        auto const& transport_properties = _process_data.getTransportProperties(
+            _element.getID(),
+            _process_variables[transport_process_id][0].get().getName());
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: I'd think this `[0]` is a component id here.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 12]
```diff
@@ -98,19 +110,19 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+          _process_variables(process_variables)
     {
         // This assertion is valid only if all nodal d.o.f. use the same shape
         // matrices.
-        // _per_process_variables.size() can represent number of nodal DOFs
+        // process_variables[0].size() can represent number of nodal DOFs
         // regardless of what scheme is adopted.
         assert(local_matrix_size ==
-               ShapeFunction::NPOINTS * _per_process_variables.size());
+               ShapeFunction::NPOINTS * _process_variables[0].size());
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: ... and this `[0]` is some process id.
It would be better, if the `[0]` accesses would be replaced with corresponding process or component ids.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 10 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 1 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Remove the commit for processing this issue which is not on propose to this pull request. I prefer to set up another pull request later on to resolve this particular issue.

**Comment by renchao-lu** [comment]
I intentionally keep the numerical setups, e.g., convergence criteria and time stepping, used in the benchmarks with staggered coupling scheme as same as used in those with monolithic coupling scheme. As you see, the calculation accuracy is inferior to that using monolithic scheme, which is a significant characteristics of staggered scheme.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 11 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: This [0] is a product of using staggered coupling scheme. Component id is the transport process id.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 12 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: It is hard to resolve...

**Comment by renchao-lu** [comment]
The changes made in the time-loop are to get the coupled solutions of previous time step.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 12 above)
Comment (TomFischer) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Why not give the correct process variable to the `LocalAssemblerData` constructor instead of a vector of vector of `ProcessVariable`?

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 13]
```diff
@@ -98,19 +111,19 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
         bool is_axially_symmetric,
         unsigned const integration_order,
         ComponentTransportProcessData const& process_data,
-        std::vector<std::reference_wrapper<ProcessVariable>> const&
-            per_process_variables)
+        std::vector<std::vector<std::reference_wrapper<ProcessVariable>>> const&
+            process_variables)
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: If there aren't any important reasons I would like to keep the previous version.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 14]
```diff
@@ -358,27 +371,329 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+    void assembleHydraulicEquation(double const t,
+                                   std::vector<double>& local_M_data,
+                                   std::vector<double>& local_K_data,
+                                   std::vector<double>& local_b_data,
+                                   LocalCoupledSolutions const& coupled_xs)
+    {
+        auto local_p = Eigen::Map<const NodalVectorType>(
+            coupled_xs.local_coupled_xs[hydraulic_process_id].data(),
+            pressure_size);
+        int const transport_process_id = 1;
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Is this `transport_process_id` only used in `assembleHydraulicEquation()`-method or is this value used somewhere else?

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 15]
```diff
@@ -358,27 +371,329 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+        }
+    }
+
+    void assembleComponentTransportEquation(
+        double const t, std::vector<double>& local_M_data,
+        std::vector<double>& local_K_data,
+        std::vector<double>& /*local_b_data*/,
+        LocalCoupledSolutions const& coupled_xs)
+    {
+        auto const transport_process_id = coupled_xs.process_id;
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Is this the same `transport_process_id` as in the above method?

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 16]
```diff
@@ -358,27 +371,329 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
         std::vector<double>& cache) const override
     {
-        auto const n_integration_points =
-            _integration_method.getNumberOfPoints();
-
         auto const indices = NumLib::getIndices(_element.getID(), dof_table);
         assert(!indices.empty());
-        auto const local_x = current_solution.get(indices);
 
-        // Assuming that fluid density always depends on the concentration of
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Is this comment obsolete now?

File: ProcessLib/ComponentTransport/ComponentTransportProcess.cpp [hunk 17]
```diff
@@ -73,14 +73,50 @@ void ComponentTransportProcess::assembleConcreteProcess(
...
+    _coupled_solutions->coupled_xs_t0.clear();
+    _coupled_solutions->coupled_xs_t0.reserve(number_of_coupled_solutions);
+    for (std::size_t i = 0; i < number_of_coupled_solutions; i++)
+    {
+        const auto& x_t0 = _xs_previous_timestep[i];
+        if (x_t0 == nullptr)
+        {
+            OGS_FATAL(
+                "Memory is not allocated for the global vector "
+                "of the solution of the previous time step for the ."
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportProcess.cpp]: The dot at the end of the line is not necessary.

File: ProcessLib/ComponentTransport/ComponentTransportProcess.cpp [hunk 18]
```diff
@@ -73,14 +73,50 @@ void ComponentTransportProcess::assembleConcreteProcess(
...
+                "staggered scheme.\n It can be done by overriding "
+                "Process::preTimestepConcreteProcess"
+                "(ref. "
+                "ComponentTransportProcess::preTimestepConcreteProcess) ");
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportProcess.cpp]: Is this message helpful for the user? What can the user do in such a situation?

File: ProcessLib/ComponentTransport/ComponentTransportProcess.h [hunk 19]
```diff
@@ -132,11 +138,16 @@ class ComponentTransportProcess final : public Process
...
 
+    void setCoupledSolutionsOfPreviousTimeStep();
+
     ComponentTransportProcessData _process_data;
 
     std::vector<std::unique_ptr<ComponentTransportLocalAssemblerInterface>>
         _local_assemblers;
 
+    /// Solutions of the previous time step
+    std::array<std::unique_ptr<GlobalVector>, 2> _xs_previous_timestep;
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportProcess.h]: What is the meaning of `_xs` in the variable name?

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 13 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Please take a look at line 539:
```
auto const& component = phase.component(
            _process_variables[transport_process_id][0].get().getName());
```
You may understand my motivation.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 12 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: For staggered coupling scheme, the information of process variables is no longer queued in a non-nesting vector container.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 14 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: used in the function members where we need to consider the dependency of fluid density on the concentration:

- assembleHydraulicEquation() at Line 403;
- getIntPtDarcyVelocity() at Line 681;
- Eigen::Vector3d getFlux() at Line 777;

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 15 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Not the same.

File: ProcessLib/HydroMechanics/Tests.cmake [hunk 20]
```diff
@@ -114,5 +114,5 @@ AddTest(
     InjectionProduction1D_Mono_pcs_0_ts_100_t_8640000.000000.vtu InjectionProduction1D_pcs_1_ts_100_t_8640000.000000.vtu epsilon_yy epsilon_yy 1e-11 1e-11
     InjectionProduction1D_Mono_pcs_0_ts_100_t_8640000.000000.vtu InjectionProduction1D_pcs_1_ts_100_t_8640000.000000.vtu sigma_yy sigma_yy 1e-11 1e-11
     InjectionProduction1D_Mono_pcs_0_ts_100_t_8640000.000000.vtu InjectionProduction1D_pcs_1_ts_100_t_8640000.000000.vtu HydraulicFlow HydraulicFlow 1e-11 0
-    InjectionProduction1D_Mono_pcs_0_ts_100_t_8640000.000000.vtu InjectionProduction1D_pcs_1_ts_100_t_8640000.000000.vtu NodalForces NodalForces 2e-7 0
+    InjectionProduction1D_Mono_pcs_0_ts_100_t_8640000.000000.vtu InjectionProduction1D_pcs_1_ts_100_t_8640000.000000.vtu NodalForces NodalForces 3e-7 0
```
Comment (TomFischer) [inline ProcessLib/HydroMechanics/Tests.cmake]: Why is this necessary?

File: ProcessLib/ComponentTransport/ComponentTransportProcess.cpp (hunk 18 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportProcess.cpp]: Worthless for users. I think we could discard the if-condition statement. What's your suggestion? @wenqing

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 16 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed.

File: ProcessLib/ComponentTransport/ComponentTransportProcess.h (hunk 19 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportProcess.h]: Simply follow the implementation in other processes... _xs indicates that we have multiple solutions.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 13 above)
Comment (TomFischer) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Okay, I see you need the process variable for the hydraulic process and process variables for the for the chemical components. Below, you can to write:
`auto const& hydraulic_process_variable = _process_variables[hydraulic_process_id][0];`
IMHO this would make the code more readable.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 12 above)
Comment (TomFischer) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: You could pick `hydraulic_process_variable` and `component_process_variables` from the `process_variables` vector.  In the monolithic implementation the process_variables could be the same. What do you think?

File: ProcessLib/HydroMechanics/Tests.cmake (hunk 20 above)
Comment (renchao-lu) [reply ProcessLib/HydroMechanics/Tests.cmake]: I understand your concerns that implementation of staggered scheme in the component transport process should not influence on other processes. I hold the same viewpoint so that a separate commit is made for this change. I think this disturbance arises from the changes in the function `solveCoupledEquationSystemsByStaggeredScheme`.

```
       if (global_coupling_iteration == 0)
        {
            int process_id = 0;
            for (auto& process_data : _per_process_data)
            {
                auto& x = *_process_solutions[process_id];
                // Copy the solution of the previous time step to a vector that
                // belongs to process. For some problems, both of the current
                // solution and the solution of the previous time step are
                // required for the coupling computation.
                process_data->process.preTimestep(x, t, dt, process_id);
                ++process_id;
            }
        }
```

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 14 above)
Comment (TomFischer) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Then this should not be a local variable. What if someone changes it only in one of those functions.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 15 above)
Comment (TomFischer) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: So both variables have the same name, but are used for different things. Which one is the 'real' transport process id?

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 11 above)
Comment (TomFischer) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: I think the intention of the code would be more clear if you would introduce somewhere above
```auto const& transport_process_variables = _process_variables[transport_process_id]```
or even
```auto const& first_transport_process_variable = _process_variables[transport_process_id][0]```.

File: ProcessLib/ComponentTransport/ComponentTransportProcess.cpp [hunk 21]
```diff
@@ -73,14 +73,50 @@ void ComponentTransportProcess::assembleConcreteProcess(
...
+    }
     // Call global assembler for each local assembly item.
     GlobalExecutor::executeSelectedMemberDereferenced(
         _global_assembler, &VectorMatrixAssembler::assemble, _local_assemblers,
-        pv.getActiveElementIDs(), dof_table, t, x, M, K, b,
+        pv.getActiveElementIDs(), dof_tables, t, x, M, K, b,
         _coupled_solutions);
 }
 
+void ComponentTransportProcess::setCoupledSolutionsOfPreviousTimeStep()
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/ComponentTransportProcess.cpp]: I am going to check the same function in HT

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 14 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed. First_transport_process_id used instead.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 13 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed. `_transport_process_variables` introduced in local assembler.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 12 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Following your suggestion, I have made changes accordingly.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 11 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 15 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed.

File: ProcessLib/ComponentTransport/ComponentTransportProcess.cpp (hunk 18 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportProcess.cpp]: Removed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 22]
```diff
@@ -201,7 +210,7 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
     }
 
     void assembleBlockMatrices(
-        int const component_id, double const t,
+        unsigned const component_id, double const t,
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: int was better.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 23]
```diff
@@ -53,22 +54,39 @@ class ComponentTransportLocalAssemblerInterface
...
-    // Pressure always first
-    static const int pressure_index = 0;
+    // Pressure
+    static const int hydraulic_process_id = 0;  // For staggered scheme
+    static const int pressure_index = 0;        // For monolithic scheme
     static const int pressure_size = ShapeFunction::NPOINTS;
-    // Size of concentration to each component
-    static const int concentration_size = ShapeFunction::NPOINTS;
+    // Concentration
+    static const int first_transport_process_id = 1;  // For staggered scheme
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: The name implies there is a second transport process, isn't it?

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 24]
```diff
@@ -358,27 +364,330 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
-        // Assuming that fluid density always depends on the concentration of
-        // the component placed at the first.
-        auto const concentration_index = 1 * ShapeFunction::NPOINTS;
-        // get local_C and local_p
-        auto const C_nodal_values = Eigen::Map<const NodalVectorType>(
-            &local_x[concentration_index], concentration_size);
-        auto const p_nodal_values = Eigen::Map<const NodalVectorType>(
-            &local_x[pressure_index], pressure_size);
+
+        if (_coupled_solutions == nullptr)  // monolithic scheme
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: _use_monolithic_scheme

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 25]
```diff
@@ -358,27 +364,330 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+            // Assuming that fluid density always depends on the concentration
+            // of the component which is placed at the uppermost.
+            auto const local_p = Eigen::Map<const NodalVectorType>(
+                &local_x[pressure_index], pressure_size);
+            auto const local_C = Eigen::Map<const NodalVectorType>(
+                &local_x[first_concentration_index], concentration_size);
+
+            return calculateIntPtDarcyVelocity(t, local_p, local_C, cache);
+        }
+        else  // staggered scheme
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: no else after return

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 26]
```diff
@@ -516,8 +845,8 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
     ComponentTransportProcessData const& _process_data;
 
     IntegrationMethod const _integration_method;
-    std::vector<std::reference_wrapper<ProcessVariable>> const&
-        _per_process_variables;
+    std::vector<std::reference_wrapper<ProcessVariable>> const
+        _transport_process_variables;
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: So, these are only the concentration variables, aren't they?
And therefore the hydraulic process variable is stored somewhere else.

File: ProcessLib/ComponentTransport/ComponentTransportProcess.cpp [hunk 27]
```diff
@@ -46,12 +46,30 @@ void ComponentTransportProcess::initializeConcreteProcess(
     unsigned const integration_order)
 {
     const int process_id = 0;
-    ProcessLib::ProcessVariable const& pv = getProcessVariables(process_id)[0];
+    ProcessLib::ProcessVariable& pv = getProcessVariables(process_id)[0];
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportProcess.cpp]: Why is the const gone?

File: ProcessLib/ComponentTransport/ComponentTransportProcess.cpp [hunk 28]
```diff
@@ -107,8 +149,58 @@ Eigen::Vector3d ComponentTransportProcess::getFlux(std::size_t const element_id,
     std::vector<GlobalIndexType> indices_cache;
     auto const r_c_indices = NumLib::getRowColumnIndices(
         element_id, *_local_to_global_index_map, indices_cache);
-    std::vector<double> local_x(x.get(r_c_indices.rows));
-    return _local_assemblers[element_id]->getFlux(p, t, local_x);
+
+    if (_coupled_solutions == nullptr)  // monolithic scheme
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportProcess.cpp]: _use_monolithic_scheme

File: ProcessLib/ComponentTransport/ComponentTransportProcess.cpp (hunk 27 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportProcess.cpp]: By mistake. Fixed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 23 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: I intend to make this variable different from another variable 'transport_process_id'. This variable is only used in calculation of fluid density which is herein assumed to be dependent upon the concentration of 1st component.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 22 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Fixed.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 24 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: _use_monolithic_scheme is not accessible in this scope, despite being my preference as well.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 25 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: I have not got your point...

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 26 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: In my opinion, picking hydraulic process variable out from the process variables as well is not necessary.

File: ProcessLib/ComponentTransport/ComponentTransportProcess.cpp (hunk 28 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportProcess.cpp]: It can be used in this scope. Fixed

**Comment by renchao-lu** [comment]
All the changes have been done accordingly. May I ask you do you have any further comment? @endJunction
